### PR TITLE
IGNITE-23685 Add local events about the HA zone topology reduce.

### PR DESCRIPTION
**Pull request by kgusakov**
https://issues.apache.org/jira/browse/IGNITE-23685

### CODE REVIEW

File: modules/distribution-zones/src/main/java/org/apache/ignite/internal/distributionzones/DistributionZoneManager.java [hunk 1]
```diff
@@ -409,7 +413,15 @@ private void onUpdatePartitionDistributionResetBusy(
             }
 
             if (partitionDistributionResetTimeoutSeconds == IMMEDIATE_TIMER_VALUE) {
-                // TODO: IGNITE-23599 Implement valid behaviour here.
+                fireEvent(
```
Comment (sanpwc) [inline modules/distribution-zones/src/main/java/org/apache/ignite/internal/distributionzones/DistributionZoneManager.java:416]: I thought that we've agreed that we won't call event in MS Watch thread, aren't we?

File: modules/distribution-zones/src/main/java/org/apache/ignite/internal/distributionzones/DistributionZoneManager.java [hunk 2]
```diff
@@ -985,22 +997,23 @@ private CompletableFuture<Void> scheduleTimers(CatalogZoneDescriptor zone, boole
...
-                    }
-
                     if (partitionReset != INFINITE_TIMER_VALUE) {
                         zonesState.get(zoneId).reschedulePartitionDistributionReset(
                                 partitionReset,
-                                // TODO: IGNITE-23599 Implement valid behaviour here.
-                                () -> {},
-                                zoneId
+                                () -> {
+                                    fireEvent(
```
Comment (sanpwc) [inline modules/distribution-zones/src/main/java/org/apache/ignite/internal/distributionzones/DistributionZoneManager.java:1004]: I'd rather hide it intro private method.

File: modules/distribution-zones/src/main/java/org/apache/ignite/internal/distributionzones/events/HaZoneTopologyUpdateEventParams.java [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+ * Parameters of {@link HaZoneTopologyUpdateEvent}.
+ */
+// TODO: https://issues.apache.org/jira/browse/IGNITE-23640 Remove this params together with the event.
+@Deprecated
+public class HaZoneTopologyUpdateEventParams implements EventParameters {
+    /** Zone id. */
+    private final int zoneId;
+
+    /**
+     * Long view of{@link org.apache.ignite.internal.hlc.HybridTimestamp}.
```
Comment (sanpwc) [inline modules/distribution-zones/src/main/java/org/apache/ignite/internal/distributionzones/events/HaZoneTopologyUpdateEventParams.java:32]: What a mix of 
`/** ... */`
 and 
```
/**
 *
 */
```

File: modules/distribution-zones/src/main/java/org/apache/ignite/internal/distributionzones/DistributionZoneManager.java (hunk 1 above)
Comment (kgusakov) [reply modules/distribution-zones/src/main/java/org/apache/ignite/internal/distributionzones/DistributionZoneManager.java:416]: Fixed

File: modules/distribution-zones/src/main/java/org/apache/ignite/internal/distributionzones/events/HaZoneTopologyUpdateEventParams.java (hunk 3 above)
Comment (kgusakov) [reply modules/distribution-zones/src/main/java/org/apache/ignite/internal/distributionzones/events/HaZoneTopologyUpdateEventParams.java:32]: Fixed

File: modules/distribution-zones/src/main/java/org/apache/ignite/internal/distributionzones/events/HaZoneTopologyUpdateEventParams.java [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+    private final long timestamp;
+
+    /**
+     * Constructor.
+     *
+     * @param causalityToken Causality token.
+     * @param zoneId Zone id.
+     * @param timestamp Long view of{@link org.apache.ignite.internal.hlc.HybridTimestamp}.
+     */
+    public HaZoneTopologyUpdateEventParams(long causalityToken, int zoneId, long timestamp) {
```
[the remark this answers is not in the document]
Comment (kgusakov) [reply modules/distribution-zones/src/main/java/org/apache/ignite/internal/distributionzones/events/HaZoneTopologyUpdateEventParams.java:43]: removed

File: modules/table/src/integrationTest/java/org/apache/ignite/internal/table/distributed/disaster/ItHighAvailablePartitionsRecoveryTest.java [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+
+        int zoneId = node.catalogManager().zone(ZONE_NAME, clock.nowLong()).id();
+        int tableId = node.catalogManager().table(TABLE_NAME, clock.nowLong()).id();
+
+        assertEquals(zoneId, request.zoneId());
+        assertEquals(tableId, request.tableId());
+        assertEquals(Set.of(0, 1), request.partitionIds());
+        assertFalse(request.manualUpdate());
+    }
+}
```
[the remark this answers is not in the document]
Comment (kgusakov) [reply modules/table/src/integrationTest/java/org/apache/ignite/internal/table/distributed/disaster/ItHighAvailablePartitionsRecoveryTest.java:101]: Test added

File: modules/distribution-zones/src/main/java/org/apache/ignite/internal/distributionzones/DistributionZoneManager.java (hunk 2 above)
Comment (kgusakov) [reply modules/distribution-zones/src/main/java/org/apache/ignite/internal/distributionzones/DistributionZoneManager.java]: Extracted to the method

File: modules/table/src/integrationTest/java/org/apache/ignite/internal/table/distributed/disaster/ItHighAvailablePartitionsRecoveryTest.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (kgusakov) [reply modules/table/src/integrationTest/java/org/apache/ignite/internal/table/distributed/disaster/ItHighAvailablePartitionsRecoveryTest.java:208]: Added

[the remark this answers is not in the document]
Comment (kgusakov) [reply modules/table/src/integrationTest/java/org/apache/ignite/internal/table/distributed/disaster/ItHighAvailablePartitionsRecoveryTest.java:208]: Added

File: modules/distribution-zones/src/main/java/org/apache/ignite/internal/distributionzones/DistributionZoneManager.java [hunk 6]
```diff
@@ -1020,6 +1009,37 @@ private CompletableFuture<Void> scheduleTimers(CatalogZoneDescriptor zone, boole
...
+     *
+     * @param revision Metastore revision.
+     * @return Appropriate metastore timestamp or -1 if revision is already compacted.
+     */
+    private long timestampByRevision(long revision) {
+        try {
+            return metaStorageManager.timestampByRevisionLocally(revision).longValue();
+        } catch (CompactedException e) {
+            if (revision > 1) {
+                LOG.warn("Unable to retrieve timestamp by revision because of meta storage compaction, [revision={}].", revision);
```
Comment (sanpwc) [inline modules/distribution-zones/src/main/java/org/apache/ignite/internal/distributionzones/DistributionZoneManager.java:1023]: Could you please expand log message a bit, by adding something like: "zone reset is skipped."?

Comment (kgusakov) [reply modules/distribution-zones/src/main/java/org/apache/ignite/internal/distributionzones/DistributionZoneManager.java:1023]: Will do in the next PR, if you don't mind

**Review by sanpwc — APPROVED** [review]

**Pull request merged**